target_chains/ethereum/contracts/forge-test/Entropy.t.sol

```diff
@@ -528,4 +546,76 @@ contract EntropyTest is Test, EntropyTestUtils {
         assertEq(providerInfo1.uri, provider1Uri);
         assertEq(providerInfo1.commitmentMetadata, provider1CommitmentMetadata);
     }
+
+    function testSetAdminByAdmin() public {
+        vm.prank(admin);
+        random.setAdmin(admin2);
+        assertEq(random.getAdmin(), admin2);
+    }
+
+    function testSetAdminByOwner() public {
+        vm.prank(owner);
+        random.setAdmin(admin2);
+        assertEq(random.getAdmin(), admin2);
+    }
+
+    function testFailSetAdminByUnauthorised() public {
+        vm.prank(admin2);
+        random.setAdmin(admin);
+    }
+
+    function testSetPythFeeByAdmin() public {
+        vm.prank(admin);
+        random.setPythFee(10);
+        assertEq(random.getPythFee(), 10);
+    }
+
+    function testSetPythFeeByOwner() public {
+        vm.prank(owner);
+        random.setPythFee(10);
+        assertEq(random.getPythFee(), 10);
+    }
+
+    function testFailSetPythFeeByUnauthorised() public {
+        vm.prank(admin2);
+        random.setPythFee(10);
+    }
+
+    function testSetDefaultProviderByOwner() public {
+        vm.prank(owner);
+        random.setDefaultProvider(provider2);
+        assertEq(random.getDefaultProvider(), provider2);
+    }
+
+    function testSetDefaultProviderByAdmin() public {
+        vm.prank(admin);
+        random.setDefaultProvider(provider2);
+        assertEq(random.getDefaultProvider(), provider2);
+    }
+
+    function testFailSetDefaultProviderByUnauthorised() public {
+        vm.prank(admin2);
+        random.setDefaultProvider(provider2);
+        assertEq(random.getDefaultProvider(), provider2);
+    }
+
+    function testUpgradeByOwner() public {
+        vm.prank(owner);
+        random.upgradeTo(address(random2));
+    }
+
+    function testFailUpgradeByAdmin() public {
+        vm.prank(admin);
+        random.upgradeTo(address(random2));
+    }
+
+    function testFailUpgradeByUnauthorised() public {
+        vm.prank(provider1);
+        random.upgradeTo(address(random2));
+    }
+
+    function testFailDifferentMagicContractUpgrade() public {
+        vm.prank(owner);
+        random.upgradeTo(address(randomDifferentMagic));
+    }
 }
```

Comment:
There can be another case that the magic function doesn't exist but it's fine. (It will revert with no reason)

Reply:
adding a comment for this